File: frontend/src/lib/spots.ts
import type { CongestionLevel } from './congestion'

export type SpotCategory = 'restaurant' | 'tourist' | 'facility'

export type Spot = {
  id: string
  name: string
  category: SpotCategory
  lat: number
  lng: number
  congestion: CongestionLevel
}

// 検索機能: 表示範囲全体に分布する飲食店・観光地・施設のピン（デモ用データ）
export const SEARCH_SPOTS: Spot[] = [
  { id: 'spot-1', name: '杉本町食堂', category: 'restaurant', lat: 34.53736, lng: 135.49599, congestion: 2 },
  { id: 'spot-2', name: '帝塚山カフェ', category: 'restaurant', lat: 34.53637, lng: 135.49995, congestion: 1 },
  { id: 'spot-3', name: '万代ラーメン', category: 'restaurant', lat: 34.53618, lng: 135.5059, congestion: 3 },
  { id: 'spot-4', name: '苅田中華料理店', category: 'restaurant', lat: 34.53745, lng: 135.5129, congestion: 2 },
  { id: 'spot-5', name: '我孫子定食屋', category: 'restaurant', lat: 34.53577, lng: 135.51781, congestion: 1 },
  { id: 'spot-6', name: '浅香ベーカリー', category: 'restaurant', lat: 34.54098, lng: 135.49436, congestion: 1 },
  { id: 'spot-7', name: '北畠居酒屋', category: 'restaurant', lat: 34.54087, lng: 135.50059, congestion: 1 },
  { id: 'spot-8', name: '山坂うどん', category: 'restaurant', lat: 34.53982, lng: 135.50747, congestion: 3 },
  { id: 'spot-9', name: '沢之町喫茶店', category: 'restaurant', lat: 34.54148, lng: 135.51211, congestion: 1 },
  { id: 'spot-10', name: '住吉焼肉店', category: 'restaurant', lat: 34.54009, lng: 135.51631, congestion: 2 },
  { id: 'spot-11', name: '東粉浜そば', category: 'restaurant', lat: 34.54448, lng: 135.49529, congestion: 1 },
  { id: 'spot-12', name: '南田辺洋食店', category: 'restaurant', lat: 34.54529, lng: 135.502, congestion: 1 },
  { id: 'spot-13', name: '清水丘食堂', category: 'restaurant', lat: 34.5437, lng: 135.5064, congestion: 3 },
  { id: 'spot-14', name: '長居カフェ', category: 'restaurant', lat: 34.54524, lng: 135.5108, congestion: 1 },
  { id: 'spot-15', name: '万代池公園', category: 'tourist', lat: 34.54456, lng: 135.5175, congestion: 2 },
  { id: 'spot-16', name: '苅田神社', category: 'tourist', lat: 34.54802, lng: 135.49649, congestion: 3 },
  { id: 'spot-17', name: '帝塚山公園', category: 'tourist', lat: 34.54949, lng: 135.5007, congestion: 4 },
  { id: 'spot-18', name: '住吉公園', category: 'tourist', lat: 34.54786, lng: 135.50533, congestion: 2 },
  { id: 'spot-19', name: '北畠展望台', category: 'tourist', lat: 34.54879, lng: 135.5122, congestion: 1 },
  { id: 'spot-20', name: '山坂緑地', category: 'tourist', lat: 34.54905, lng: 135.51844, congestion: 4 },
  { id: 'spot-21', name: '浅香史跡公園', category: 'tourist', lat: 34.55174, lng: 135.495, congestion: 2 },
  { id: 'spot-22', name: '我孫子庭園', category: 'tourist', lat: 34.55342, lng: 135.4999, congestion: 2 },
  { id: 'spot-23', name: '東粉浜緑地', category: 'tourist', lat: 34.55226, lng: 135.5069, congestion: 1 },
  { id: 'spot-24', name: '長居公園', category: 'tourist', lat: 34.55229, lng: 135.51286, congestion: 4 },
  { id: 'spot-25', name: '杉本町図書館', category: 'facility', lat: 34.5534, lng: 135.51681, congestion: 3 },
  { id: 'spot-26', name: '沢之町市民センター', category: 'facility', lat: 34.53473, lng: 135.4925, congestion: 2 },
  { id: 'spot-27', name: '南田辺体育館', category: 'facility', lat: 34.55408, lng: 135.52007, congestion: 3 },
  { id: 'spot-28', name: '清水丘コミュニティホール', category: 'facility', lat: 34.53526, lng: 135.51424, congestion: 4 },
  { id: 'spot-29', name: '苅田資料館', category: 'facility', lat: 34.55339, lng: 135.49865, congestion: 3 },
  { id: 'spot-30', name: '帝塚山ホール', category: 'facility', lat: 34.5455, lng: 135.49213, congestion: 4 },
]

File: backend/app/routers/spots.py
import asyncio

import httpx
from fastapi import APIRouter, Depends, HTTPException, Query
from pydantic import BaseModel
from sqlalchemy import and_, func, or_, select
from sqlalchemy.ext.asyncio import AsyncSession

from app.database import get_db
from app.dependencies import get_current_user
from app.models.store import Review, Store, StoreTag, Tag
from app.schemas.spot import SpotResponse, SpotSearchResponse
from app.services.supabase_auth import SupabaseAuthResult

router = APIRouter(prefix="/spots", tags=["spots"])

_geocode_lock = asyncio.Lock()
_last_geocode_time: float = 0.0


class GeocodeResult(BaseModel):
    latitude: float
    longitude: float
    displayName: str


class SpotDetailResponse(BaseModel):
    spotId: str
    name: str
    description: str
    category: str
    latitude: float
    longitude: float
    congestionStatus: str
    reviewRating: float


class SpotReviewItem(BaseModel):
    reviewId: str
    userName: str
    rating: float
    content: str
    photoUrl: str | None


class SpotReviewsResponse(BaseModel):
    reviews: list[SpotReviewItem]

# フロントのジャンル ID → tags.name
GENRE_TAG_NAMES: dict[str, str] = {
    "cafe": "カフェ",
    "restaurant": "レストラン",
    "temple": "寺院・神社・城",
    "park": "公園",
    "museum": "美術館・博物館・ミュージアム",
}


def _congestion_label(level: float | None) -> str:
    """crowrd_level（0.0=空き 〜 1.0=混雑）を表示用ラベルに変換する。"""
    if level is None:
        return "不明"
    if level < 0.25:
        return "空いている"
    if level < 0.5:
        return "少し空き"
    if level < 0.75:
        return "少し混雑"
    return "混雑"


def _congestion_filter(crowrd_level, values: list[str]):
    ranges = {
        "empty": and_(crowrd_level >= 0.0, crowrd_level < 0.25),
        "few_empty": and_(crowrd_level >= 0.25, crowrd_level < 0.5),
        "few_crowded": and_(crowrd_level >= 0.5, crowrd_level < 0.75),
        "crowded": and_(crowrd_level >= 0.75, crowrd_level <= 1.0),
    }
    conditions = [ranges[v] for v in values if v in ranges]
    return or_(*conditions) if conditions else None


def _review_filter(avg_rating, values: list[str]):
    ranges = {
        "4.1-5.0": avg_rating > 4.0,
        "3.1-4.0": and_(avg_rating > 3.0, avg_rating <= 4.0),
        "2.1-3.0": and_(avg_rating > 2.0, avg_rating <= 3.0),
        "2.0-below": avg_rating <= 2.0,
    }
    conditions = [ranges[v] for v in values if v in ranges]
    return or_(*conditions) if conditions else None


@router.get("/search", response_model=SpotSearchResponse)
async def search_spots(
    congestion: list[str] = Query(default=[]),
    genre: list[str] = Query(default=[]),
    review: list[str] = Query(default=[]),
    db: AsyncSession = Depends(get_db),
    _user: SupabaseAuthResult = Depends(get_current_user),
) -> SpotSearchResponse:
    """フィルター条件に合うスポット（店舗）一覧を取得する。"""

    # 店舗ごとの平均レビュー（レビュー無しは 0.0）
    avg_rating = (
        select(func.coalesce(func.avg(Review.rating), 0.0))
        .where(Review.store == Store.store_id)
        .scalar_subquery()
    )
    # 店舗の代表カテゴリ（タグ名を 1 件）
    category = (
        select(Tag.name)
        .join(StoreTag, StoreTag.tag == Tag.tag_id)
        .where(StoreTag.store == Store.store_id)
        .order_by(Tag.name)
        .limit(1)
        .scalar_subquery()
    )

    stmt = select(
        Store.store_id,
        Store.name,
        Store.lat,
        Store.lon,
        Store.crowrd_level,
        category.label("category"),
        avg_rating.label("avg_rating"),
    )

    congestion_cond = _congestion_filter(Store.crowrd_level, congestion)
    if congestion_cond is not None:
        stmt = stmt.where(congestion_cond)

    if genre:
        names = [GENRE_TAG_NAMES[g] for g in genre if g in GENRE_TAG_NAMES]
        if names:
            genre_exists = (
                select(1)
                .select_from(StoreTag)
                .join(Tag, StoreTag.tag == Tag.tag_id)
                .where(StoreTag.store == Store.store_id, Tag.name.in_(names))
                .exists()
            )
            stmt = stmt.where(genre_exists)
        else:
            return SpotSearchResponse(spots=[])

    review_cond = _review_filter(avg_rating, review)
    if review_cond is not None:
        stmt = stmt.where(review_cond)

    result = await db.execute(stmt)
    rows = result.all()

    spots = [
        SpotResponse(
            spotId=str(row.store_id),
            name=row.name,
            category=row.category or "",
            latitude=row.lat if row.lat is not None else 0.0,
            longitude=row.lon if row.lon is not None else 0.0,
            congestionStatus=_congestion_label(row.crowrd_level),
            reviewRating=round(float(row.avg_rating), 1),
        )
        for row in rows
    ]
    return SpotSearchResponse(spots=spots)


@router.get("/geocode", response_model=GeocodeResult)
async def geocode_address(
    q: str = Query(..., min_length=1),
    _user: SupabaseAuthResult = Depends(get_current_user),
) -> GeocodeResult:
    """住所・地名を緯度経度に変換する（Nominatim 経由）。Nominatim の利用規約に従い 1req/s 制限。"""
    global _last_geocode_time
    import time

    async with _geocode_lock:
        now = time.monotonic()
        wait = 1.0 - (now - _last_geocode_time)
        if wait > 0:
            await asyncio.sleep(wait)
        _last_geocode_time = time.monotonic()

    url = "https://nominatim.openstreetmap.org/search"
    params = {"q": q, "format": "json", "limit": 1}
    headers = {"User-Agent": "58hack-in-OMU/1.0 (demo)"}
    async with httpx.AsyncClient(timeout=10.0) as client:
        res = await client.get(url, params=params, headers=headers)
        res.raise_for_status()
    data = res.json()
    if not data:
        raise HTTPException(status_code=404, detail="場所が見つかりませんでした")
    item = data[0]
    return GeocodeResult(
        latitude=float(item["lat"]),
        longitude=float(item["lon"]),
        displayName=item.get("display_name", q),
    )


@router.get("/{spot_id}", response_model=SpotDetailResponse)
async def get_spot(
    spot_id: str,
    db: AsyncSession = Depends(get_db),
    _user: SupabaseAuthResult = Depends(get_current_user),
) -> SpotDetailResponse:
    """スポット（店舗）の詳細情報を取得する。"""
    avg_rating = (
        select(func.coalesce(func.avg(Review.rating), 0.0))
        .where(Review.store == Store.store_id)
        .scalar_subquery()
    )
    category = (
        select(Tag.name)
        .join(StoreTag, StoreTag.tag == Tag.tag_id)
        .where(StoreTag.store == Store.store_id)
        .order_by(Tag.name)
        .limit(1)
        .scalar_subquery()
    )
    result = await db.execute(
        select(
            Store.store_id,
            Store.name,
            Store.description,
            Store.lat,
            Store.lon,
            Store.crowrd_level,
            category.label("category"),
            avg_rating.label("avg_rating"),
        ).where(Store.store_id == spot_id)
    )
    row = result.first()
    if row is None:
        raise HTTPException(status_code=404, detail="スポットが見つかりません")
    return SpotDetailResponse(
        spotId=str(row.store_id),
        name=row.name,
        description=row.description,
        category=row.category or "",
        latitude=row.lat if row.lat is not None else 0.0,
        longitude=row.lon if row.lon is not None else 0.0,
        congestionStatus=_congestion_label(row.crowrd_level),
        reviewRating=round(float(row.avg_rating), 1),
    )


@router.get("/{spot_id}/reviews", response_model=SpotReviewsResponse)
async def get_spot_reviews(
    spot_id: str,
    db: AsyncSession = Depends(get_db),
    _user: SupabaseAuthResult = Depends(get_current_user),
) -> SpotReviewsResponse:
    """指定スポットのレビュー一覧を取得する。"""
    from app.models.user import User

    result = await db.execute(
        select(
            Review.review_id,
            Review.rating,
            Review.content,
            Review.photo_url,
            User.name.label("user_name"),
        )
        .join(User, User.user_id == Review.post_user)
        .where(Review.store == spot_id)
        .order_by(Review.review_id.desc())
    )
    rows = result.all()
    return SpotReviewsResponse(
        reviews=[
            SpotReviewItem(
                reviewId=str(row.review_id),
                userName=row.user_name,
                rating=row.rating or 0.0,
                content=row.content,
                photoUrl=row.photo_url,
            )
            for row in rows
        ]
    )
